### PR TITLE
Add new vsphere node driver

### PR DESCRIPTION
**Pull request by Rohit-PX**
###  Why do we need this PR?
* Adding ability to interact with the ESXi VMs via the _SSH_ node drivers
* Using the govimomi API library to interact with the Vsphere SDK

Signed-off-by: Rohit-PX <rohit@portworx.com>

### CODE REVIEW

File: drivers/node/vsphere/vsphere.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+	// Make future calls local to this datacenter
+	f.SetDatacenter(dc)
+
+	// Find virtual machines in datacenter
+	vms, err := f.VirtualMachineList(v.ctx, "*")
+	if err != nil {
+		return fmt.Errorf("Failed to find any virtual machines on %s: %v", v.vsphereHostIP, err)
+	}
+
+	nodes := node.GetWorkerNodes()
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:149]: Change to have all nodes including master node in the vmMap

File: drivers/node/vsphere/vsphere.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+// RebootVM reboots vsphere VM
+func (v *vsphere) RebootVM(n node.Node) error {
+	if _, ok := vmMap[n.Name]; !ok {
+		return fmt.Errorf("Could not fetch VM for node: %s", n.Name)
+	}
+
+	vm := vmMap[n.Name]
+
+	fmt.Printf("\nRebooting VM: %s  ", vm.Name())
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:170]: logrus instead of fmt

File: drivers/node/vsphere/vsphere.go [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+			}
+		}
+	}
+
+	return nil
+}
+
+// ShutdownVM shutsdown the vsphere VM
+func (v *vsphere) ShutdownVM(n node.Node) error {
+	return nil
```
Comment (Rohit-PX) [inline drivers/node/vsphere/vsphere.go:225]: Implement ShutdownVM as well

File: drivers/scheduler/k8s/specs/mysql/pxd/px-storage-class.yaml [hunk 4]
```diff
@@ -19,7 +19,7 @@ parameters:
   fs: "ext4"
   block_size: "512b"
```
Comment (Rohit-PX) [inline drivers/scheduler/k8s/specs/mysql/pxd/px-storage-class.yaml:20]: Delete this file from the PR

File: tests/reboot/reboot_test.go [hunk 5]
```diff
@@ -56,47 +56,45 @@ var _ = Describe("{RebootOneNode}", func() {
 			// Reboot node and check driver status
 			Step(fmt.Sprintf("reboot node one at a time from the node(s): %v", nodesToReboot), func() {
 				for _, n := range nodesToReboot {
-					if n.IsStorageDriverInstalled {
```
Comment (Rohit-PX) [inline tests/reboot/reboot_test.go:59]: This change is not required revert this change

File: drivers/node/node.go [hunk 6]
```diff
@@ -152,6 +152,15 @@ type Driver interface {
...
 	GetZones() ([]string, error)
+
+	// PowerOnVM powers VM
+	PowerOnVM(node Node) error
+
+	// PowerOffVM powers VM
+	PowerOffVM(node Node) error
+
+	// RebootVM powers VM
+	RebootVM(node Node) error
```
Comment (Rohit-PX) [inline drivers/node/node.go:163]: Use the existing RebootNode methods and override them in the vsphere driver implementation

File: drivers/node/vsphere/vsphere.go (hunk 1 above)
Comment (Rohit-PX) [reply drivers/node/vsphere/vsphere.go:149]: Fixed.

File: drivers/node/vsphere/vsphere.go (hunk 2 above)
Comment (Rohit-PX) [reply drivers/node/vsphere/vsphere.go:170]: Fixed in all places.

File: drivers/node/node.go (hunk 6 above)
Comment (Rohit-PX) [reply drivers/node/node.go]: Fixed

File: drivers/node/vsphere/vsphere.go (hunk 1 above)
Comment (Rohit-PX) [reply drivers/node/vsphere/vsphere.go]: Fixed

File: drivers/node/vsphere/vsphere.go [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+	vsphereIP    = "VSPHERE_HOST_IP"
+)
+
+const (
+	// DefaultUsername is the default username used for ssh operations
+	DefaultUsername = "root"
+	// VMReadyTimeout Timeout for checking VM power state
+	VMReadyTimeout = 3 * time.Minute
+	// VMRetryInterval interval for retry when checking VM power state
+	VMRetryInterval = 5 * time.Second
```
Comment (mborodin-px) [inline drivers/node/vsphere/vsphere.go:39]: rename it with the same name as above VMReadyRetryInterval

File: drivers/node/vsphere/vsphere.go [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+
+const (
+	vsphereUname = "VSPHERE_USER"
+	vspherePwd   = "VSPHERE_PWD"
+	vsphereIP    = "VSPHERE_HOST_IP"
+)
+
+const (
+	// DefaultUsername is the default username used for ssh operations
+	DefaultUsername = "root"
```
Comment (mborodin-px) [inline drivers/node/vsphere/vsphere.go:35]: rename it to DefaultVsphereUsername as this is a username for vSphere

File: drivers/node/vsphere/vsphere.go [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+func (v *vsphere) Init() error {
+	logrus.Infof("Using the vsphere node driver")
+	v.SSH.Init()
+
+	username := os.Getenv(vsphereUname)
+	if len(username) == 0 {
+		v.vsphereUsername = DefaultUsername
+	} else {
+		v.vsphereUsername = username
+	}
```
Comment (mborodin-px) [inline drivers/node/vsphere/vsphere.go:70]: this construction will be more readable if you use like this:
```
	v.vsphereUsername = DefaultUsername
	username := os.Getenv(vsphereUname)
	if username != "" {
		v.vsphereUsername = username
	}
```

File: drivers/node/vsphere/vsphere.go [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+				Cause: fmt.Sprintf("Failed to test connection to VM: %s Current Status: %v, error: %v", vm.Name(), powerState, err),
+			}
+		}
+		return nil, false, nil
+	}
+	if _, err := task.DoRetryWithTimeout(t, VMReadyTimeout, VMRetryInterval); err != nil {
+		return err
+	}
+	return err
+}
```
Comment (mborodin-px) [inline drivers/node/vsphere/vsphere.go:111]: The reason for this method is not just check if node is up, but also it is running. The VM could be in boot process and can't execute any commands. Add the execution of simple command like date or hostname as we use in ssh driver.

File: drivers/node/vsphere/vsphere.go [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+	return nil
+}
+
+// RebootVM reboots vsphere VM
+func (v *vsphere) RebootNode(n node.Node, options node.RebootNodeOpts) error {
+	if _, ok := vmMap[n.Name]; !ok {
+		return fmt.Errorf("Could not fetch VM for node: %s", n.Name)
+	}
+
+	vm := vmMap[n.Name]
```
Comment (mborodin-px) [inline drivers/node/vsphere/vsphere.go:168]: I would implement getVmByName and get a VM from vSphere in runtime, then you don't need to keep vmMap

File: drivers/node/vsphere/vsphere.go [hunk 12]
```diff
@@ -0,0 +1,248 @@
...
+			Cause: fmt.Sprintf("failed to reboot VM %s. cause %v", vm.Name(), err),
+		}
+	}
+	return nil
+}
+
+// PowerOnVM powers on the VM if not already on
+func (v *vsphere) PowerOnVM(n node.Node) error {
+	var err error
+	vm := vmMap[n.Name]
```
Comment (mborodin-px) [inline drivers/node/vsphere/vsphere.go:184]: see comment above

File: drivers/node/vsphere/vsphere.go [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+
+	logrus.Infof("Powering on VM: %s  ", vm.Name())
+	tsk, err := vm.PowerOn(v.ctx)
+	if err != nil {
+		logrus.Errorf("Failed to power off %s: %v", vm.Name(), err)
+	} else {
```
Comment (mborodin-px) [inline drivers/node/vsphere/vsphere.go:190]: redundant else here

File: drivers/node/vsphere/vsphere.go (hunk 8 above)
Comment (Rohit-PX) [reply drivers/node/vsphere/vsphere.go:35]: Since this is already in the vsphere driver, I think we need to call it `DefaultVsphereUsername` same convention is being followed in the ssh driver

File: drivers/node/vsphere/vsphere.go (hunk 9 above)
Comment (Rohit-PX) [reply drivers/node/vsphere/vsphere.go:70]: makes sense, fixed.

File: drivers/node/vsphere/vsphere.go (hunk 7 above)
Comment (Rohit-PX) [reply drivers/node/vsphere/vsphere.go:39]: Fixed.

File: drivers/node/ssh/vsphere/vsphere.go [hunk 14]
```diff
@@ -0,0 +1,175 @@
...
+			tsk, err := vm.PowerOn(ctx)
+			if err != nil {
+				fmt.Printf("Failed to power off %s: %v", vm.Name(), err)
+			} else {
+				if _, err := tsk.WaitForResult(ctx); err != nil {
+					fmt.Printf("failed to wait for VM %s to be powered off. cause %v", vm.Name(), err)
+				}
+			}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/node/ssh/vsphere/vsphere.go]: This was from the old code as I was still rebasing, fixed now.

File: drivers/node/ssh/vsphere/vsphere.go [hunk 15]
```diff
@@ -0,0 +1,175 @@
...
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	c, err := govmomi.NewClient(ctx, u, true)
+	if err != nil {
+		return fmt.Errorf("logging in error: %s", err.Error())
+	}
+
+	fmt.Printf("Log in successful. \nList of VMs on %s:\n", HostIP)
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/node/ssh/vsphere/vsphere.go]: This was from the old code as I was still rebasing, fixed now.

File: drivers/node/ssh/vsphere/vsphere.go [hunk 16]
```diff
@@ -0,0 +1,175 @@
...
+	return err
+}
+
+func (v *Vsphere) connect(n node.Node) error {
+
+	HostIP := "70.0.0.197"
+	//protocol := "https://"
+
+	//login := "https://administrator@vsphere.local:Password1@70.0.75.173/sdk"
+	login := fmt.Sprintf("https://root:P@$$w0rd!@%s/sdk", HostIP)
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/node/ssh/vsphere/vsphere.go]: This was from the old code as I was still rebasing, fixed now.

File: drivers/node/ssh/vsphere/vsphere.go [hunk 17]
```diff
@@ -0,0 +1,175 @@
+package vsphere
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+	"text/tabwriter"
+	//	"strings"
+	//	"time"
```
[the remark this answers is not in the document]
Comment (Rohit-PX) [reply drivers/node/ssh/vsphere/vsphere.go]: This was from the old code as I was still rebasing, fixed now.

File: drivers/node/vsphere/vsphere.go (hunk 10 above)
Comment (Rohit-PX) [reply drivers/node/vsphere/vsphere.go:111]: Added execution of hostname command once it's verified that VM is powered on.

File: drivers/node/vsphere/vsphere.go (hunk 11 above)
Comment (Rohit-PX) [reply drivers/node/vsphere/vsphere.go:178]: I would say, it makes sense to have an in-memory object that contains only the VMs being used in this test rather than querying the vSphere every time. Currently there is a similar data structure nodeMap being used for the same reason.

File: drivers/node/vsphere/vsphere.go (hunk 12 above)
Comment (Rohit-PX) [reply drivers/node/vsphere/vsphere.go:194]: Replied above.

File: drivers/node/vsphere/vsphere.go (hunk 13 above)
Comment (Rohit-PX) [reply drivers/node/vsphere/vsphere.go:200]: yes, I need to return an error 
```if err != nil```

**Review by thiguetta — APPROVED** [review]
LGTM
